### PR TITLE
Make array Range indexer expression an rvalue

### LINKED ISSUE #31889: Range indexer on array should probably not be ref-returning
The following works:

```C#
using System;
public class C {
    public void M(int[] arr) {
        ref int[] x = ref arr[0..2];
    }
}
```

But I'm not sure it should.

### PR DESCRIPTION
**Pull request by agocke**
Fixes #31889

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 1]
```diff
@@ -383,8 +383,27 @@ internal bool CheckValueKind(SyntaxNode node, BoundExpression expr, BindValueKin
...
+                    {
+                        if (RequiresRefAssignableVariable(valueKind))
+                        {
+                            Error(diagnostics, ErrorCode.ERR_RefLocalOrParamExpected, node);
+                            return false;
+                        }
+
+                        var boundAccess = (BoundArrayAccess)expr;
+                        if (boundAccess.Indices.Length == 1 &&
+                            boundAccess.Indices[0].Type.Equals(Compilation.GetWellKnownType(WellKnownType.System_Range)))
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Shouldn't we be using `TypeSymbol.Equals` here?

File: src/Compilers/CSharp/Test/Semantic/Semantics/IndexAndRangeTests.cs [hunk 2]
```diff
@@ -2,12 +2,33 @@
...
 
 namespace Microsoft.CodeAnalysis.CSharp.UnitTests
 {
     public class IndexAndRangeTests : CompilingTestBase
     {
+        [Fact]
+        [WorkItem(31889, "https://github.com/dotnet/roslyn/issues/31889")]
+        public void ArrayRangeIllegalRef()
+        {
+            var comp = CreateCompilationWithIndexAndRange(@"
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/IndexAndRangeTests.cs]: Should have a test to verify the behavior when Range is missing.

File: src/Compilers/CSharp/Test/Semantic/Semantics/IndexAndRangeTests.cs [hunk 3]
```diff
@@ -2,12 +2,33 @@
...
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (4,27): error CS1510: A ref or out value must be an assignable variable
+                //         ref int[] x = ref arr[0..2];
+                Diagnostic(ErrorCode.ERR_RefLvalueExpected, "arr[0..2]").WithLocation(4, 27),
+                // (5,20): error CS8156: An expression cannot be used in this context because it may not be passed or returned by reference
+                //         return ref arr[0..2];
+                Diagnostic(ErrorCode.ERR_RefReturnLvalueExpected, "arr[0..2]").WithLocation(5, 20));
+        }
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/IndexAndRangeTests.cs]: What is the behavior for Index?

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs (hunk 1 above)
Comment (agocke) [reply src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: Isn't that what this is? There's a static method that just calls the instance method. Are you worried about if the LHS type is null?

**Comment by agocke** [comment]
@jaredpar Anything else?

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs (hunk 1 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs]: I was pretty sure @jcouv was moving everything to be `TypeSymbol.Equals` and removing the instance on entirely. Or at least he's pushing every equality check to use the enum that specifies what type of equality is meant to be done here (null considering or not for example). 

---
In reply to: [248118094](https://github.com/dotnet/roslyn/pull/32449#discussion_r248118094) [](ancestors = 248118094)

File: src/Compilers/CSharp/Test/Semantic/Semantics/IndexAndRangeTests.cs [hunk 4]
```diff
@@ -2,12 +2,101 @@
...
+
+        [Fact]
+        [WorkItem(31889, "https://github.com/dotnet/roslyn/issues/31889")]
+        public void ArrayRangeIllegalRefNoRange()
+        {
+            var comp = CreateCompilationWithIndex(@"
+public class C {
+    public void M(int[] arr) {
+        ref int[] x = ref arr[0..2];
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/IndexAndRangeTests.cs]: Shouldn't we have a test that verifies that we can, or cannot, take the ref of an array access when the indexer is of type `Index`?

**Comment by jaredpar** [comment]
Done with review pass (iteration 2).

**Comment by agocke** [comment]
I figured out what you were talking about, and found a bug. Updated.

**Pull request merged**